servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/DelayedSubscription.java

```diff
@@ -18,84 +18,104 @@
-        // Temporarily wrap in a ConcurrentSubscription to prevent concurrent invocation between this thread and
-        // a thread which may be interacting with this class's Subscription API.
-        final Subscription concurrentSubscription = wrap(delayedSubscription);
-        if (!currentUpdater.compareAndSet(this, null, concurrentSubscription)) {
-            delayedSubscription.cancel();
-        } else {
-            tryDrainRequested(concurrentSubscription);
-
-            // Unwrap the concurrent subscription because there will be no more concurrency.
-            currentUpdater.compareAndSet(this, concurrentSubscription, delayedSubscription);
+    public boolean delayedSubscription(Subscription delayedSubscription) {
+        requireNonNull(delayedSubscription);
+        for (;;) {
+            final long prevRequested = requested;
+            if (prevRequested <= INVALID_GREATEST_VALUE) {
+                delayedSubscription.cancel();
+                return false;
+            } else if (requestedUpdater.compareAndSet(this, prevRequested, SUBSCRIPTION_SETTING)) {
+                if (prevRequested != 0) {
+                    delayedSubscription.request(prevRequested);
+                }
+                // Set the subscription before CAS to make it visible to the thread interacting with the Subscription.
+                // The Subscription thread won't use the state unless the CAS to SUBSCRIPTION_SET is successful, so
+                // there will be no concurrency introduced by this operation.
+                subscription = delayedSubscription;
+                if (requestedUpdater.compareAndSet(this, SUBSCRIPTION_SETTING, SUBSCRIPTION_SET)) {
+                    return true;
+                }
+            }
         }
     }
 
     @Override
     public void request(long n) {
-        Subscription current = this.current;
-        if (current != null) {
-            current.request(n);
-        } else {
-            if (isRequestNValid(n)) {
-                requestedUpdater.accumulateAndGet(this, n, FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-            } else {
-                // Although 0 is invalid we use it to signify that we have drained the pending request count,
-                // so in this case we use MIN_VALUE so we can still pass through an invalid number only once.
-                requested = n == 0 ? MIN_VALUE : n;
-            }
-            current = this.current;
-            if (current != null) {
-                tryDrainRequested(current);
+        for (;;) {
+            final long prevRequested = requested;
+            if (prevRequested == SUBSCRIPTION_SET) {
+                assert subscription != null;
+                subscription.request(n);
+                break;
+            } else if (prevRequested == SUBSCRIPTION_SETTING) {
+                if (requestedUpdater.compareAndSet(this, SUBSCRIPTION_SETTING, addRequestN(0, n))) {
```

Comment:
Considering the case:
1. Initial `requested == 0`
2. Thread 1 invokes `delayedSubscription`, does `requested = SUBSCRIPTION_SETTING`.
3. Thread 2 does `request(0)`, does `requested = -1`.
3. Thread 1 goes to the next for loop iteration and does `delayedSubscription.request(-1)` because `prevRequested == -1`.
While both values `0` and `-1` are illegal according to rule 3.9 it doesn't sound right that we change the original `n` value from `0` to `-1`. I'm not sure if this is acceptable or not.
If not, consider having another state that will represent `request(0)`: `REQUESTED_ZERO = SUBSCRIPTION_CANCEL_PENDING + 1`.

Reply:
we are overloading the `long` domain to serve multiple purposes:

1. concurrency control
2. accumulation of demand
3. representing invalid demand
4. representing cancel

The trade-off is we lose some fidelity (e.g. minimum of 3 invalid `request(n)` values) on the "representing invalid demand" in order to represent these state transition. Ideally we preserve the original `n` for debuggability but we use a similar approach in other places to reduce state and consolidate atomic operations on a single field.

We can preserve `0` via some additional mapping/unmapping if we think that is more likely to be used than some other negative number (doesn't seem unreasonable), but also don't think it is a hard requirement based upon the spec.